root_numpy/src/tree.pyx

```diff
@@ -2,43 +2,22 @@ include "converters.pyx"
 
 
 def list_trees(fname):
-    cdef TFile* rfile = Open(fname, 'read')
-    if rfile == NULL:
-        raise IOError("cannot read {0}".format(fname))
-    cdef TList* keys = rfile.GetListOfKeys()
-    if keys == NULL:
-        raise IOError("unable to get keys in {0}".format(fname))
-    ret = dict()
-    cdef int nkeys = keys.GetEntries()
-    cdef TKey* key
-    for i in range(nkeys):
-        key = <TKey*> keys.At(i)
-        clsname = str(key.GetClassName())
-        if clsname == 'TTree' or clsname == 'TNtuple':
-            ret[str(key.GetName())] = None
-    rfile.Close()
-    del rfile
-    return list(ret.keys())
+    objects = list_objects(fname)
+    return [key for key, val in objects.items() if val == 'TTree' or val == 'TNtuple']
 
 
 def list_directories(fname):
-    cdef TFile* rfile = Open(fname, 'read')
+    objects = list_objects(fname)
+    return [key for key, val in objects.items() if val == 'TDirectoryFile']
+
+def list_objects(fname):
```

Comment:
I think list_objects should also take an optional list of class names and only populate the dictionary with keys that match those names. Then in list_trees above you could have ``objects = list_objects(fname, types=['TTree', 'TNtuple'])``.